base/reduce.jl

```diff
@@ -204,6 +197,32 @@ foldr(op, itr) = mapfoldr(identity, op, itr)
     end
 end
 
+# The @simd transformation is only valid in limited situations
+struct SIMDableFunction{f}; end
+(::SIMDableFunction{f})(args...) where {f} = f(args...)
+simdable(f::Union{map(typeof, (+, *, &, |, add_sum, mul_prod, -, /, ^, identity))...}) = SIMDableFunction{f}()
```

Comment:
I am not sure that sure that this is meaningful. The question is when is a `function` not valid to be executed as simd?
When it has side-effects? `@simd` is an advise to LLVM that gives it permissions to perform some limited transformations  and encourages it to be more lenient, but LLVM still has to proof that the simd transformation is valid and beneficial.

Reply:
Yes, this is a poor name.  I don't really want to be building a trait-system here — I wanted to just use a hard-coded `Union{}` — but due to bootstrapping issues not all the functions I wanted to whitelist were available.  Of course we can not guarantee that these functions won't have an invalid method on them.

Really, the problem is that `@simd` is non-local.  It affects any function that gets inlined — even passed user functions.  We have no idea if the particular method is going to have observable side-effects that break the SIMD guarantees.  This is an attempt at a white-list to recoup some of the performance in common cases — but perhaps we shouldn't even try to do that.

> LLVM still has to proof that the simd transformation is valid and beneficial

My understanding is that `@simd` is promising that the transformation is valid — allowing LLVM to do things that it cannot prove.

Reply:
> allowing LLVM to do things that it cannot prove.

Essentially true, but the list above is largely redundant with things that LLVM should be able to prove (since it's approximately just a subset list of known pure functions), and, beyond that, is incorrect in general (there's no guarantee – or expectation – that my implementation of any of those functions is actually pure).

Reply:
> My understanding is that `@simd` is promising that the transformation is valid

It defines that _certain_ transforms are valid, but we implement that list. Currently it contains:
 - primitive ops that contribute towards the computation of a floating point computation reduction may be reordered (reassociated / commuted)
 - all memory writes are independent of all other memory writes and reads that are performed inside the loop (e.g. all memory that is read from is readonly, and no location is written to twice)

Reply:
Ah, great, that is extremely helpful.  So if we disable the memory independence assertion (https://github.com/JuliaLang/julia/commit/b68efd34e7e3827c1c63d92d3551b57ab90ec378), then is it correct that one safely could annotate `@simd` over unknown generic functions?  My read is yes: this simply adds extra handling for _reduction variables_ — which, by definition, must be lexically visible to the `@simd` author.

Memory dependencies, on the other hand, might _not_ be lexically visible to the `@simd` author. Doing something like #19658 could perhaps allow us to recoup some of the lost perf in the future in a much safer and precise way.

Reply:
That's my understanding too. Although there's some non-locality in the definition of what constitutes "computation of the reduction", since we just examine the use/def graph after optimization. It is possible to make a closure which obscures the reduction variable lexically. For example:
```julia
result = Ref(0.0)
for i = 1:10
    result[] += i
end
```
```julia
result = Ref(0.0)
update(i) = (result[] += i)
for i = 1:10
    update(i)
end
```
This pattern appears, for example, in the `Stateful` iterator.